packages/zk-eddsa-event-ticket-pcd/src/ZKEdDSAEventTicketPCD.ts

```diff
@@ -0,0 +1,538 @@
+      .getTrapdoor()
+      .toString(),
+
+    // Valid event ID list
+    validEventIds: snarkInputForValidEventIds(validEventIdsInput),
+    checkValidEventIds: checkValidEventIds ? "1" : "0",
+
+    // Security features
+    externalNullifier:
+      externalNullifer || STATIC_TICKET_PCD_NULLIFIER.toString(),
+    revealNullifierHash: externalNullifer ? "1" : "0",
+    watermark: watermark.toString()
+  };
+}
+
+function claimFromProofResult(
+  ticketPCD: EdDSATicketPCD,
+  publicSignals: string[],
+  validEventIds: string[] | undefined,
+  externalNullifer: string | undefined,
+  watermark: bigint
+): ZKEdDSAEventTicketPCDClaim {
+  const partialTicket: Partial<ITicketData> = {};
+  if (!babyJubIsNegativeOne(publicSignals[0])) {
+    partialTicket.ticketId = decStringToBigIntToUuid(publicSignals[0]);
+  }
+  if (!babyJubIsNegativeOne(publicSignals[1])) {
+    partialTicket.eventId = decStringToBigIntToUuid(publicSignals[1]);
+  }
+  if (!babyJubIsNegativeOne(publicSignals[2])) {
+    partialTicket.productId = decStringToBigIntToUuid(publicSignals[2]);
+  }
+  if (!babyJubIsNegativeOne(publicSignals[3])) {
+    partialTicket.timestampConsumed = parseInt(publicSignals[3]);
+  }
+  if (!babyJubIsNegativeOne(publicSignals[4])) {
+    partialTicket.timestampSigned = parseInt(publicSignals[4]);
+  }
+  if (!babyJubIsNegativeOne(publicSignals[5])) {
+    partialTicket.attendeeSemaphoreId = publicSignals[5];
+  }
+  if (!babyJubIsNegativeOne(publicSignals[6])) {
+    partialTicket.isConsumed = publicSignals[6] !== "0";
+  }
+  if (!babyJubIsNegativeOne(publicSignals[7])) {
+    partialTicket.isRevoked = publicSignals[7] !== "0";
+  }
+
+  const claim: ZKEdDSAEventTicketPCDClaim = {
+    partialTicket,
+    watermark: watermark.toString(),
+    signer: ticketPCD.proof.eddsaPCD.claim.publicKey
+  };
+
+  if (validEventIds !== undefined) {
+    claim.validEventIds = validEventIds;
+  }
+
+  if (externalNullifer !== undefined) {
+    claim.nullifierHash = publicSignals[8];
+    claim.externalNullifier = externalNullifer;
+  }
+
+  return claim;
+}
+
+/**
+ * Creates a new ZKEdDSAEventTicketPCD.
+ */
+export async function prove(
+  args: ZKEdDSAEventTicketPCDArgs
+): Promise<ZKEdDSAEventTicketPCD> {
+  if (!initArgs) {
+    throw new Error("Cannot make proof: init has not been called yet");
+  }
+  await ensureInitialized();
+
+  const { ticketPCD, identityPCD, fieldsToReveal, watermark } =
+    await checkProveInputs(args);
+
+  const snarkInput = snarkInputForProof(
+    ticketPCD,
+    identityPCD,
+    fieldsToReveal,
+    args.validEventIds?.value,
+    args.externalNullifier?.value,
+    watermark
+  );
+
+  const { proof, publicSignals } = await groth16.fullProve(
+    snarkInput,
+    initArgs.wasmFilePath,
+    initArgs.zkeyFilePath
+  );
+
+  const claim = claimFromProofResult(
+    ticketPCD,
+    publicSignals,
+    args.validEventIds?.value,
+    args.externalNullifier?.value,
+    watermark
+  );
```

Comment:
this is pretty confusing to me - why does the claim include information that is derived from the proof?

Reply:
> in the abstract case the proof is a validated computation, so it's expected that you'd make claims about its outputs

yes, i think @ichub's confusion is because the word "proof" is sort of overloaded. there's "proof" in the snarkjs sense and then there's "proof" in the PCD sense. snarkjs "proof" = PCD "proof" + snarkjs "public signals." snarkjs public signals generally correspond to the PCD claim though.

> inputs should go directly from PCD args to claim, while outputs should go from proof to claim.

i might be too sleepy to be saying something that makes sense here but i don't think that in general inputs should go from PCD args to claim. consider the case where some of your PCD args are meant to be private (i.e. one of your PCD args is a semaphore public/private keypair; you wouldn't want those in the claim). i think typically it is the case that a snarkjs proof's `publicSignals` are the claim. (there are cases in my `ZKEdDSATicketPCDClaim` where i appear to be pulling something from `args` into claim, but i could just as easily have pulled from `publicSignals`, since those args are public inputs to the SNARK that will simply be copied into `publicSignals`)

> In the case of outputs, I think taking them from the proof makes the most sense since otherwise you'd be duplicating the proof's verified computation

i agree with this